src/CattyTests/Formula/FormulaTest.swift

```diff
@@ -0,0 +1,126 @@
+        let screenSize = Util.screenSize(true)
+        formulaManager = FormulaManager(sceneSize: screenSize)
+        interpreter = formulaManager
+    }
+
+    func testIsSingularNumberFormula() {
+        var formula = Formula(integer: 1)!
+        XCTAssertTrue(formula.isSingularNumber())
+
+        formula = Formula(double: 1.0)!
+        XCTAssertTrue(formula.isSingularNumber())
+
+        formula = Formula(float: 1.0)!
+        XCTAssertTrue(formula.isSingularNumber())
+
+        formula = Formula(string: "1")!
+        XCTAssertTrue(formula.isSingularNumber())
+
+        formula = Formula(string: "1.0")!
+        XCTAssertTrue(formula.isSingularNumber())
+
+        formula = Formula(double: 1.1)!
+        XCTAssertFalse(formula.isSingularNumber())
+
+        let internTokenList = NSMutableArray()
+        var token = InternToken(type: TOKEN_TYPE_OPERATOR, andValue: MinusOperator.tag)!
+        var tokenNumber = InternToken(type: TOKEN_TYPE_NUMBER, andValue: "1")!
+        internTokenList.add(token)
+        internTokenList.add(tokenNumber)
+
+        var internParser = InternFormulaParser(tokens: (internTokenList as! [InternToken]), andFormulaManager: formulaManager)
+
+        var parseTree = internParser!.parseFormula(for: nil)
+        formula = Formula(formulaElement: parseTree)
+
+        XCTAssertNotNil(parseTree)
+        XCTAssertEqual(-1, interpreter.interpretDouble(formula, for: SpriteObject()))
+        internTokenList.removeAllObjects()
+
+        formula = Formula(formulaElement: parseTree)
+        XCTAssertFalse(formula.isSingularNumber())
+
+        token = InternToken(type: TOKEN_TYPE_OPERATOR, andValue: MinusOperator.tag)
+        tokenNumber = InternToken(type: TOKEN_TYPE_NUMBER, andValue: "1.0")
+        internTokenList.add(token)
+        internTokenList.add(tokenNumber)
+
+        internParser = InternFormulaParser(tokens: (internTokenList as! [InternToken]), andFormulaManager: formulaManager)
+        parseTree = internParser!.parseFormula(for: nil)
+
+        XCTAssertNotNil(parseTree)
+        XCTAssertEqual(-1, interpreter.interpretDouble(formula, for: SpriteObject()))
+        internTokenList.removeAllObjects()
+
+        formula = Formula(formulaElement: parseTree)
+        XCTAssertFalse(formula.isSingularNumber())
+
+        token = InternToken(type: TOKEN_TYPE_OPERATOR, andValue: MinusOperator.tag)
+        tokenNumber = InternToken(type: TOKEN_TYPE_NUMBER, andValue: "1.0")
+        let secondToken = InternToken(type: TOKEN_TYPE_OPERATOR, andValue: MinusOperator.tag)!
+        let secondNumber = InternToken(type: TOKEN_TYPE_NUMBER, andValue: "1.0")!
+        internTokenList.add(token)
+        internTokenList.add(tokenNumber)
+        internTokenList.add(secondToken)
+        internTokenList.add(secondNumber)
+
+        internParser = InternFormulaParser(tokens: (internTokenList as! [InternToken]), andFormulaManager: formulaManager)
+        parseTree = internParser!.parseFormula(for: nil)
```

Comment:
Please try to reduce the overall amount of force unwraps in this file, thank you!

Reply:
Thank you for the tip!
I think I fixed all the force unwraps.